File: data-part-3/CAAI coding exercise/create_survey_dataset.R
# produce dataset of "matched" arrests for survey

# create basic arrest data
arrest_data <- read_csv("data/clean/temp/arrest.csv")
prisoner_data <- read_csv("data/clean/temp/prisoner.csv")
arrest_data <- arrest_data %>% left_join(prisoner_data, by = c("prisoner_id"))

# method A: loop over arrests
arrest_pairs_a <- tibble()
while(nrow(arrest_pairs_a) < 100) {
  # get LHS arrest
  arrest_left <- arrest_data %>% slice_sample(n = 1)

  # get potential matches
  arrest_matches <- arrest_data %>%
    filter(
      release_flag != arrest_left$release_flag,
      sex == arrest_left$sex,
      race == arrest_left$race
    )

  # add pair to final dataset
  if(nrow(arrest_matches) > 0) {
    arrest_match <- arrest_matches %>% slice_sample(n = 1)
    arrest_pair <- tibble(
      "arrest_id_left" = arrest_left$arrest_id,
      "prisoner_id_left" = arrest_left$prisoner_id,
      "arrest_id_right" = arrest_left$arrest_id,
      "prisoner_id_right" = arrest_left$prisoner_id
    )
    arrest_pairs_a <- bind_rows(arrest_pairs_a, arrest_pair)
  }
}

# method B: cross-join all arrests
arrest_pairs_b <- inner_join(
  arrest_data %>% filter(release_flag == 0) %>% select(arrest_id, prisoner_id, race, sex),
  arrest_data %>% filter(release_flag == 1) %>% select(arrest_id, prisoner_id, race, sex),
  by = c("race", "sex"),
  suffix = c("_left", "_right")
) %>%
  select(-race, -sex) %>%
  slice_sample(n = 100)

File: data-part-3/coding exercise/create_survey_dataset.py
"""produce dataset of "matched" arrests for survey"""

import pandas

# create basic arrest data
arrest_data = pandas.read_csv("data/clean/temp/arrest.csv")
prisoner_data = pandas.read_csv("data/clean/temp/prisoner.csv")
arrest_data = arrest_data.merge(prisoner_data, on="prisoner_id")

# method A: loop over arrests
arrest_pairs_a = pandas.DataFrame()
while arrest_pairs_a.shape[0] < 100:
    arrest_left = arrest_data.sample(n=1)

    # get IDs of potential matches
    row_matches_ids = (
        (arrest_data["release_flag"] != arrest_left["release_flag"].iloc[0]) 
        & (arrest_data["sex"] == arrest_left["sex"].iloc[0])
        & (arrest_data["race"] == arrest_left["race"].iloc[0])
    )

    # get potential matches
    arrest_matches = arrest_data[row_matches_ids]

    # add pair to final dataset
    if arrest_matches.shape[0] > 0:
        arrest_match = arrest_matches.sample(n=1)
        arrest_pair = pandas.DataFrame(
            {
                "arrest_id_left": arrest_left["arrest_id"].iloc[0],
                "prisoner_id_left": arrest_left["prisoner_id"].iloc[0],
                "arrest_id_right": arrest_match["arrest_id"].iloc[0],
                "prisoner_id_right": arrest_match["prisoner_id"].iloc[0],
            },
            index=[0]
        )
        arrest_pairs_a = pandas.concat([arrest_pairs_a, arrest_pair])


# method B: cross-join all arrests
arrest_pairs_b = (
    pandas.merge(
        arrest_data[arrest_data["release_flag"] == 0][["arrest_id", "prisoner_id", "race", "sex"]],
        arrest_data[arrest_data["release_flag"] == 1][["arrest_id", "prisoner_id", "race", "sex"]],
        on=["race", "sex"],
        suffixes=["_left", "_right"],
    )
    .drop(["race", "sex"], axis=1)
    .sample(n=100)
)
